clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -508,9 +508,30 @@ private void processAssignmentReceived(ConsumerGroupHeartbeatResponseData.Assign
      */
     private void replaceTargetAssignmentWithNewAssignment(
             ConsumerGroupHeartbeatResponseData.Assignment assignment) {
-        currentTargetAssignment.clear();
+
+        // Return if the same as current assignment; comparison without creating a new collection
+        if (currentTargetAssignment != null) {
+            // check if the new assignment is different from the current target assignment
+            if (currentTargetAssignment.partitions.size() == assignment.topicPartitions().size() &&
+                assignment.topicPartitions().stream().allMatch(
+                    tp -> currentTargetAssignment.partitions.containsKey(tp.topicId()) &&
+                        currentTargetAssignment.partitions.get(tp.topicId()).size() == tp.partitions().size() &&
+                        currentTargetAssignment.partitions.get(tp.topicId()).containsAll(tp.partitions()))) {
+                return;
+            }
+        }
+
+        // Bump local epoch and replace assignment
+        long nextLocalEpoch;
+        if (currentTargetAssignment == null) {
+            nextLocalEpoch = 0;
+        } else {
+            nextLocalEpoch = currentTargetAssignment.localEpoch + 1;
+        }
```

Comment:
Is there a reason why we need to compute epochs on the client here? The server is the one bumping the epochs whenever it computes a new assignment for a member. I was expecting that we just keep a `LocalAssignment` that will contain the partitions and epoch the broker sent on the `ConsumerGroupHeartbeatResponseData`

Reply:
I think the server is bumping the target member epoch when any assignment for any member changes, while the local epoch is only bumped if the assignment for this specific member changes. So we should get fewer reconciliations.

Reply:
I was wrong thinking that we could reuse the server epoch here, but it's not truly representing new assignments for the member, and we could end up receiving multiple new assignments in the same server epoch. So ok with this local epoch approach, best way to keep track of "versions" of the assignment received.